operator/cmd/mesh/uninstall.go

```diff
@@ -0,0 +1,179 @@
+	opts := &helmreconciler.Options{DryRun: rootArgs.dryRun, Log: l, ProgressLog: progress.NewLog()}
+	h, err := helmreconciler.NewHelmReconciler(clt, restConfig, nil, opts)
+	if err != nil {
+		return fmt.Errorf("failed to create reconciler: %v", err)
+	}
+
+	// If only revision flag is set, we would prune resources by the revision label.
+	// Otherwise we would merge the revision flag and the filename flag and delete resources by generated manifests.
+	if uiArgs.revision != "" && uiArgs.filename == "" {
+		objectsList, tp, err := h.GetPrunedResourcesList(uiArgs.revision)
+		if err != nil {
+			return err
+		}
+		if err := preCheckWarnings(cmd, rootArgs, uiArgs, uiArgs.revision, tp, l); err != nil {
+			return fmt.Errorf("failed to do preuninstall check: %v", err)
+		}
+
+		if err := h.PruneControlPlaneByRevision(uiArgs.revision, objectsList); err != nil {
+			return fmt.Errorf("failed to prune control plane resources by revision: %v", err)
+		}
+	}
+	uiArgs.set = append(uiArgs.set, fmt.Sprintf("revision=%s", uiArgs.revision))
+	manifestMap, iops, err := manifest.GenManifests([]string{uiArgs.filename}, applyFlagAliases(uiArgs.set, uiArgs.manifestsPath, uiArgs.revision), uiArgs.force, restConfig, l)
+	if err != nil {
+		return err
+	}
+	if iops == nil {
+		return fmt.Errorf("istioOperatorSpec is nil")
+	}
+	if err := preCheckWarnings(cmd, rootArgs, uiArgs, iops.Revision, nil, l); err != nil {
+		return fmt.Errorf("failed to do preuninstall check: %v", err)
+	}
+	cpManifests := manifestMap[name.PilotComponentName]
+	if err := h.DeleteControlPlaneByManifests(strings.Join(cpManifests, "---"), iops.Revision); err != nil {
+		return fmt.Errorf("failed to delete control plane by manifests: %v", err)
+	}
+	opts.ProgressLog.SetState(progress.StateComplete)
+	return nil
+}
+
+// preCheckWarnings checks possible breaking changes and issue warnings to users, it checks the following:
+// 1. checks proxies still pointing to the target control plane revision.
+// 2. lists to be pruned resources if user uninstall by --revision flag.
+func preCheckWarnings(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs,
+	rev string, resourcesList []string, l *clog.ConsoleLogger) error {
+	pids, err := proxyinfo.GetIDsFromProxyInfo(uiArgs.kubeConfigPath, uiArgs.context, rev, uiArgs.istioNamespace)
+	if err != nil {
+		return err
+	}
+	needConfirmation := false
+	var message string
+	if resourcesList != nil {
+		needConfirmation = true
+		message += fmt.Sprintf("To be pruned resources from the cluster: %s\n",
+			strings.Join(resourcesList, ","))
+	}
+	if len(pids) != 0 {
+		needConfirmation = true
+		message += fmt.Sprintf("There are still proxies pointing to the control plane revision %s:\n%s."+
+			" If you proceed with the uninstall, these proxies will become detached from any control plane"+
+			" and will not function correctly.. ",
+			uiArgs.revision, strings.Join(pids, " \n"))
+	}
+	if uiArgs.skipConfirmation || uiArgs.force || rootArgs.dryRun {
```

Comment:
sorry to keep on about this but it still doesn't look correct to me. we should be skipping confirmation only when that flag is selected, not for force or dry-run right? there's nothing preventing users from setting all three flags if they want the command to run without interaction in dry-run.